### PR TITLE
Move game interface to Arena

### PR DESCRIPTION
**Pull request by oliverlee**

### CODE REVIEW

**Comment by oliverlee** [comment]
I want to re-enable tests before this is merged.

File: dominion/src/dominion/arena.rs [hunk 1]
```diff
@@ -1,40 +1,515 @@
...
     }
 
     fn start_game(&mut self) {
-        for p in self.players() {
+        for p in self.players.iter_mut() {
             p.cleanup();
         }
     }
+
+    fn get_player(&self, player_id: usize) -> Result<&Player> {
```
Comment (mickvangelderen) [inline dominion/src/dominion/arena.rs]: `s/get_player/player/g`
consider adding
`current_player` and `current_player_mut` which use turn.player_id

File: dominion/src/dominion/arena.rs [hunk 2]
```diff
@@ -1,40 +1,515 @@
...
+                        }
+                    }
+                    None => Err(Error::InvalidCardChoice),
+                },
+                _ => Err(Error::InvalidCardLocation),
+            },
+        }
+    }
+
+    pub fn check_hand(&self, player_id: usize) -> Result<&CardVec> {
```
Comment (mickvangelderen) [inline dominion/src/dominion/arena.rs]: `s/check_hand/hand/g`

File: dominion/src/dominion/arena.rs [hunk 3]
```diff
@@ -1,40 +1,515 @@
...
+        Ok(&self.get_player(player_id)?.hand)
     }
 
-    pub fn players(&mut self) -> &mut Vec<Player> {
-        &mut self.players
+    pub fn check_discard_pile(&self, player_id: usize) -> Result<&CardVec> {
+        Ok(&self.get_player(player_id)?.discard_pile)
```
Comment (mickvangelderen) [inline dominion/src/dominion/arena.rs]:
```suggestion
        self.player(player_id).map(|player| &player.discard_pile)
```

**Review by mickvangelderen — CHANGES_REQUESTED** [review]
If you don't expect people to pass an invalid player id, it is reasonable to panic in methods that accept it. You can then provide a `try_*` version of the function that returns a result. The `try_` function can be used if the input is not to be trusted and an invalid player id is to be expected.

get and check don't mean much to me and I have to agree with Kevlin Henney that they just make code harder to read.

you can make functions pub(crate) to only expose then to the crate to facilitate testing and accesing internals. This allows you to provide convenience function without having to expose them in your public api.

File: dominion/src/dominion/arena.rs (hunk 2 above)
Comment (oliverlee) [reply dominion/src/dominion/arena.rs]: done

File: dominion/src/dominion/arena.rs (hunk 1 above)
Comment (oliverlee) [reply dominion/src/dominion/arena.rs]: Done and done. However, I'm not sure how we will use `current_player` and `current_player_mut` right now.

File: dominion/src/dominion/arena.rs (hunk 3 above)
Comment (oliverlee) [reply dominion/src/dominion/arena.rs]: Done

**Comment by oliverlee** [comment]
> If you don't expect people to pass an invalid player id, it is reasonable to panic in methods that accept it. You can then provide a try_* version of the function that returns a result. The try_ function can be used if the input is not to be trusted and an invalid player id is to be expected.

Makes sense but I suggest we do that later as this may change a lot at the moment.

> get and check don't mean much to me and I have to agree with Kevlin Henney that they just make code harder to read.

I think this is a good idea. Thanks for point it out.

**Pull request merged**